src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractSyntacticSingleKeywordRecommender.cs

```diff
@@ -37,21 +34,21 @@ protected virtual Task<bool> IsValidContextAsync(int position, CSharpSyntaxConte
 
         protected virtual bool IsValidContext(int position, CSharpSyntaxContext context, CancellationToken cancellationToken) => false;
 
-        public async Task<IEnumerable<RecommendedKeyword>> RecommendKeywordsAsync(
+        public async Task<ImmutableArray<RecommendedKeyword>> RecommendKeywordsAsync(
             int position,
             CSharpSyntaxContext context,
             CancellationToken cancellationToken)
         {
             var syntaxKind = await RecommendKeywordAsync(position, context, cancellationToken).ConfigureAwait(false);
             if (syntaxKind.HasValue)
             {
-                return SpecializedCollections.SingletonEnumerable(
+                return ImmutableArray.Create(
                     new RecommendedKeyword(SyntaxFacts.GetText(syntaxKind.Value),
                         shouldFormatOnCommit: ShouldFormatOnCommit,
                         matchPriority: ShouldPreselect(context, cancellationToken) ? SymbolMatchPriority.Keyword : MatchPriority.Default));
             }
 
-            return null;
+            return ImmutableArray<RecommendedKeyword>.Empty;
```

Comment:
⚠️ This path is now allocating a `Task<TResult>`, where it wasn't before. Is it called often?

Reply:
This no longer allocates tasks at all.

--

nearly all the allocations in keyword completin can likely be entirely removed.  i'm considering doing that, but it's a lot more mechanical work.